File: sdks/java/src/test/java/to/renamed/sdk/ErrorsTest.java
package to.renamed.sdk;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for error handling and exception mapping.
 */
class ErrorsTest {

    @Nested
    @DisplayName("RenamedError Base Class Tests")
    class RenamedErrorTests {

        @Test
        @DisplayName("Creates error with message and code")
        void createsErrorWithMessageAndCode() {
            RenamedError error = new RenamedError("Something went wrong", "GENERIC_ERROR");

            assertEquals("Something went wrong", error.getMessage());
            assertEquals("GENERIC_ERROR", error.getCode());
            assertEquals(0, error.getStatusCode());
            assertNull(error.getDetails());
        }

        @Test
        @DisplayName("Creates error with all properties")
        void createsErrorWithAllProperties() {
            Map<String, Object> details = Map.of("field", "value", "count", 42);

            RenamedError error = new RenamedError("Detailed error", "DETAILED_ERROR", 500, details);

            assertEquals("Detailed error", error.getMessage());
            assertEquals("DETAILED_ERROR", error.getCode());
            assertEquals(500, error.getStatusCode());
            assertNotNull(error.getDetails());
            assertEquals("value", error.getDetails().get("field"));
            assertEquals(42, error.getDetails().get("count"));
        }

        @Test
        @DisplayName("Creates error with cause")
        void createsErrorWithCause() {
            Exception cause = new RuntimeException("Root cause");
            RenamedError error = new RenamedError("Wrapped error", "WRAPPED_ERROR", cause);

            assertEquals("Wrapped error", error.getMessage());
            assertEquals("WRAPPED_ERROR", error.getCode());
            assertEquals(0, error.getStatusCode());
            assertSame(cause, error.getCause());
        }

        @Test
        @DisplayName("toString with status code")
        void toStringWithStatusCode() {
            RenamedError error = new RenamedError("Server error", "SERVER_ERROR", 500, null);

            String result = error.toString();

            assertTrue(result.contains("SERVER_ERROR"));
            assertTrue(result.contains("500"));
            assertTrue(result.contains("Server error"));
        }

        @Test
        @DisplayName("toString without status code")
        void toStringWithoutStatusCode() {
            RenamedError error = new RenamedError("Network failed", "NETWORK_ERROR");

            String result = error.toString();

            assertTrue(result.contains("NETWORK_ERROR"));
            assertTrue(result.contains("Network failed"));
            assertFalse(result.contains("status"));
        }
    }

    @Nested
    @DisplayName("RenamedError.fromHttpStatus Factory Tests")
    class FromHttpStatusTests {

        @Test
        @DisplayName("Returns AuthenticationError for 401")
        void returnsAuthenticationErrorFor401() {
            RenamedError error = RenamedError.fromHttpStatus(401, "Unauthorized", null);

            assertInstanceOf(AuthenticationError.class, error);
            assertEquals("Unauthorized", error.getMessage());
            assertEquals("AUTHENTICATION_ERROR", error.getCode());
            assertEquals(401, error.getStatusCode());
        }

        @Test
        @DisplayName("Returns InsufficientCreditsError for 402")
        void returnsInsufficientCreditsErrorFor402() {
            RenamedError error = RenamedError.fromHttpStatus(402, "Payment Required", null);

            assertInstanceOf(InsufficientCreditsError.class, error);
            assertEquals("Payment Required", error.getMessage());
            assertEquals("INSUFFICIENT_CREDITS", error.getCode());
            assertEquals(402, error.getStatusCode());
        }

        @Test
        @DisplayName("Returns ValidationError for 400")
        void returnsValidationErrorFor400() {
            Map<String, Object> payload = Map.of("error", "Invalid file format");

            RenamedError error = RenamedError.fromHttpStatus(400, "Bad Request", payload);

            assertInstanceOf(ValidationError.class, error);
            assertEquals("Invalid file format", error.getMessage());
            assertEquals("VALIDATION_ERROR", error.getCode());
            assertEquals(400, error.getStatusCode());
        }

        @Test
        @DisplayName("Returns ValidationError for 422")
        void returnsValidationErrorFor422() {
            Map<String, Object> payload = Map.of("error", "Unprocessable entity");

            RenamedError error = RenamedError.fromHttpStatus(422, "Unprocessable Entity", payload);

            assertInstanceOf(ValidationError.class, error);
            assertEquals("Unprocessable entity", error.getMessage());
        }

        @Test
        @DisplayName("Returns RateLimitError for 429")
        void returnsRateLimitErrorFor429() {
            Map<String, Object> payload = new HashMap<>();
            payload.put("error", "Too many requests");
            payload.put("retryAfter", 60);

            RenamedError error = RenamedError.fromHttpStatus(429, "Too Many Requests", payload);

            assertInstanceOf(RateLimitError.class, error);
            assertEquals("Too many requests", error.getMessage());
            assertEquals("RATE_LIMIT_ERROR", error.getCode());
            assertEquals(429, error.getStatusCode());
            assertEquals(60, ((RateLimitError) error).getRetryAfter());
        }

        @Test
        @DisplayName("Returns RateLimitError with zero retry for missing retryAfter")
        void returnsRateLimitErrorWithZeroRetryForMissingRetryAfter() {
            Map<String, Object> payload = Map.of("error", "Rate limited");

            RenamedError error = RenamedError.fromHttpStatus(429, "Too Many Requests", payload);

            assertInstanceOf(RateLimitError.class, error);
            assertEquals(0, ((RateLimitError) error).getRetryAfter());
        }

        @Test
        @DisplayName("Returns generic RenamedError for 500")
        void returnsGenericRenamedErrorFor500() {
            RenamedError error = RenamedError.fromHttpStatus(500, "Internal Server Error", null);

            assertEquals(RenamedError.class, error.getClass());
            assertEquals("Internal Server Error", error.getMessage());
            assertEquals("API_ERROR", error.getCode());
            assertEquals(500, error.getStatusCode());
        }

        @Test
        @DisplayName("Returns generic RenamedError for 503")
        void returnsGenericRenamedErrorFor503() {
            RenamedError error = RenamedError.fromHttpStatus(503, "Service Unavailable", null);

            assertEquals(RenamedError.class, error.getClass());
            assertEquals("API_ERROR", error.getCode());
            assertEquals(503, error.getStatusCode());
        }

        @Test
        @DisplayName("Extracts error message from payload")
        void extractsErrorMessageFromPayload() {
            Map<String, Object> payload = Map.of("error", "Custom error message");

            RenamedError error = RenamedError.fromHttpStatus(500, "Server Error", payload);

            assertEquals("Custom error message", error.getMessage());
        }

        @Test
        @DisplayName("Falls back to status text when payload has no error")
        void fallsBackToStatusTextWhenPayloadHasNoError() {
            Map<String, Object> payload = Map.of("someOtherField", "value");

            RenamedError error = RenamedError.fromHttpStatus(500, "Server Error", payload);

            assertEquals("Server Error", error.getMessage());
        }

        @Test
        @DisplayName("Handles null payload")
        void handlesNullPayload() {
            RenamedError error = RenamedError.fromHttpStatus(404, "Not Found", null);

            assertEquals("Not Found", error.getMessage());
        }
    }

    @Nested
    @DisplayName("AuthenticationError Tests")
    class AuthenticationErrorTests {

        @Test
        @DisplayName("Creates with default message")
        void createsWithDefaultMessage() {
            AuthenticationError error = new AuthenticationError();

            assertEquals("Invalid or missing API key", error.getMessage());
            assertEquals("AUTHENTICATION_ERROR", error.getCode());
            assertEquals(401, error.getStatusCode());
        }

        @Test
        @DisplayName("Creates with custom message")
        void createsWithCustomMessage() {
            AuthenticationError error = new AuthenticationError("API key expired");

            assertEquals("API key expired", error.getMessage());
            assertEquals("AUTHENTICATION_ERROR", error.getCode());
            assertEquals(401, error.getStatusCode());
        }

        @Test
        @DisplayName("Falls back to default message for null")
        void fallsBackToDefaultMessageForNull() {
            AuthenticationError error = new AuthenticationError(null);

            assertEquals("Invalid or missing API key", error.getMessage());
        }

        @Test
        @DisplayName("Falls back to default message for empty string")
        void fallsBackToDefaultMessageForEmptyString() {
            AuthenticationError error = new AuthenticationError("");

            assertEquals("Invalid or missing API key", error.getMessage());
        }

        @Test
        @DisplayName("Is a RenamedError")
        void isARenamedError() {
            AuthenticationError error = new AuthenticationError();

            assertInstanceOf(RenamedError.class, error);
            assertInstanceOf(RuntimeException.class, error);
        }
    }

    @Nested
    @DisplayName("ValidationError Tests")
    class ValidationErrorTests {

        @Test
        @DisplayName("Creates with message only")
        void createsWithMessageOnly() {
            ValidationError error = new ValidationError("File too large");

            assertEquals("File too large", error.getMessage());
            assertEquals("VALIDATION_ERROR", error.getCode());
            assertEquals(400, error.getStatusCode());
            assertNull(error.getDetails());
        }

        @Test
        @DisplayName("Creates with message and details")
        void createsWithMessageAndDetails() {
            Map<String, Object> details = Map.of(
                "maxSize", 10485760,
                "actualSize", 52428800
            );

            ValidationError error = new ValidationError("File too large", details);

            assertEquals("File too large", error.getMessage());
            assertNotNull(error.getDetails());
            assertEquals(10485760, error.getDetails().get("maxSize"));
            assertEquals(52428800, error.getDetails().get("actualSize"));
        }

        @Test
        @DisplayName("Is a RenamedError")
        void isARenamedError() {
            ValidationError error = new ValidationError("Invalid format");

            assertInstanceOf(RenamedError.class, error);
        }
    }

    @Nested
    @DisplayName("InsufficientCreditsError Tests")
    class InsufficientCreditsErrorTests {

        @Test
        @DisplayName("Creates with default message")
        void createsWithDefaultMessage() {
            InsufficientCreditsError error = new InsufficientCreditsError();

            assertEquals("Insufficient credits", error.getMessage());
            assertEquals("INSUFFICIENT_CREDITS", error.getCode());
            assertEquals(402, error.getStatusCode());
        }

        @Test
        @DisplayName("Creates with custom message")
        void createsWithCustomMessage() {
            InsufficientCreditsError error = new InsufficientCreditsError("You need 10 credits, but have 0");

            assertEquals("You need 10 credits, but have 0", error.getMessage());
            assertEquals("INSUFFICIENT_CREDITS", error.getCode());
        }

        @Test
        @DisplayName("Falls back to default message for null")
        void fallsBackToDefaultMessageForNull() {
            InsufficientCreditsError error = new InsufficientCreditsError(null);

            assertEquals("Insufficient credits", error.getMessage());
        }

        @Test
        @DisplayName("Falls back to default message for empty string")
        void fallsBackToDefaultMessageForEmptyString() {
            InsufficientCreditsError error = new InsufficientCreditsError("");

            assertEquals("Insufficient credits", error.getMessage());
        }
    }

    @Nested
    @DisplayName("RateLimitError Tests")
    class RateLimitErrorTests {

        @Test
        @DisplayName("Creates with default message")
        void createsWithDefaultMessage() {
            RateLimitError error = new RateLimitError();

            assertEquals("Rate limit exceeded", error.getMessage());
            assertEquals("RATE_LIMIT_ERROR", error.getCode());
            assertEquals(429, error.getStatusCode());
            assertEquals(0, error.getRetryAfter());
        }

        @Test
        @DisplayName("Creates with custom message and retry time")
        void createsWithCustomMessageAndRetryTime() {
            RateLimitError error = new RateLimitError("Slow down", 30);

            assertEquals("Slow down", error.getMessage());
            assertEquals(30, error.getRetryAfter());
        }

        @Test
        @DisplayName("Falls back to default message for null")
        void fallsBackToDefaultMessageForNull() {
            RateLimitError error = new RateLimitError(null, 45);

            assertEquals("Rate limit exceeded", error.getMessage());
            assertEquals(45, error.getRetryAfter());
        }

        @Test
        @DisplayName("Handles zero retry after")
        void handlesZeroRetryAfter() {
            RateLimitError error = new RateLimitError("Too fast", 0);

            assertEquals(0, error.getRetryAfter());
        }
    }

    @Nested
    @DisplayName("NetworkError Tests")
    class NetworkErrorTests {

        @Test
        @DisplayName("Creates with default message")
        void createsWithDefaultMessage() {
            NetworkError error = new NetworkError();

            assertEquals("Network request failed", error.getMessage());
            assertEquals("NETWORK_ERROR", error.getCode());
            assertEquals(0, error.getStatusCode());
        }

        @Test
        @DisplayName("Creates with custom message")
        void createsWithCustomMessage() {
            NetworkError error = new NetworkError("Connection refused");

            assertEquals("Connection refused", error.getMessage());
            assertEquals("NETWORK_ERROR", error.getCode());
        }

        @Test
        @DisplayName("Creates with message and cause")
        void createsWithMessageAndCause() {
            Exception cause = new java.net.ConnectException("Host unreachable");
            NetworkError error = new NetworkError("Failed to connect", cause);

            assertEquals("Failed to connect", error.getMessage());
            assertSame(cause, error.getCause());
        }

        @Test
        @DisplayName("Falls back to default message for null")
        void fallsBackToDefaultMessageForNull() {
            NetworkError error = new NetworkError(null);

            assertEquals("Network request failed", error.getMessage());
        }

        @Test
        @DisplayName("Falls back to default message for empty string")
        void fallsBackToDefaultMessageForEmptyString() {
            NetworkError error = new NetworkError("");

            assertEquals("Network request failed", error.getMessage());
        }

        @Test
        @DisplayName("Falls back to default message with cause for null message")
        void fallsBackToDefaultMessageWithCauseForNullMessage() {
            Exception cause = new RuntimeException("Original");
            NetworkError error = new NetworkError(null, cause);

            assertEquals("Network request failed", error.getMessage());
            assertSame(cause, error.getCause());
        }
    }

    @Nested
    @DisplayName("JobStatus Enum Tests")
    class JobStatusTests {

        @Test
        @DisplayName("Values have correct string representations")
        void valuesHaveCorrectStringRepresentations() {
            assertEquals("pending", JobStatus.PENDING.getValue());
            assertEquals("processing", JobStatus.PROCESSING.getValue());
            assertEquals("completed", JobStatus.COMPLETED.getValue());
            assertEquals("failed", JobStatus.FAILED.getValue());
        }

        @Test
        @DisplayName("fromValue parses correctly")
        void fromValueParsesCorrectly() {
            assertEquals(JobStatus.PENDING, JobStatus.fromValue("pending"));
            assertEquals(JobStatus.PROCESSING, JobStatus.fromValue("processing"));
            assertEquals(JobStatus.COMPLETED, JobStatus.fromValue("completed"));
            assertEquals(JobStatus.FAILED, JobStatus.fromValue("failed"));
        }

        @Test
        @DisplayName("fromValue is case insensitive")
        void fromValueIsCaseInsensitive() {
            assertEquals(JobStatus.PENDING, JobStatus.fromValue("PENDING"));
            assertEquals(JobStatus.COMPLETED, JobStatus.fromValue("Completed"));
            assertEquals(JobStatus.FAILED, JobStatus.fromValue("FAILED"));
        }

        @Test
        @DisplayName("fromValue returns null for unknown value")
        void fromValueReturnsNullForUnknownValue() {
            assertNull(JobStatus.fromValue("unknown"));
            assertNull(JobStatus.fromValue("running"));
            assertNull(JobStatus.fromValue(""));
        }

        @Test
        @DisplayName("fromValue returns null for null input")
        void fromValueReturnsNullForNullInput() {
            assertNull(JobStatus.fromValue(null));
        }

        @Test
        @DisplayName("toString returns value")
        void toStringReturnsValue() {
            assertEquals("pending", JobStatus.PENDING.toString());
            assertEquals("completed", JobStatus.COMPLETED.toString());
        }
    }

    @Nested
    @DisplayName("Error Inheritance Tests")
    class ErrorInheritanceTests {

        @Test
        @DisplayName("All errors extend RenamedError")
        void allErrorsExtendRenamedError() {
            assertInstanceOf(RenamedError.class, new AuthenticationError());
            assertInstanceOf(RenamedError.class, new ValidationError("test"));
            assertInstanceOf(RenamedError.class, new InsufficientCreditsError());
            assertInstanceOf(RenamedError.class, new RateLimitError());
            assertInstanceOf(RenamedError.class, new NetworkError());
        }

        @Test
        @DisplayName("All errors are RuntimeExceptions")
        void allErrorsAreRuntimeExceptions() {
            assertInstanceOf(RuntimeException.class, new AuthenticationError());
            assertInstanceOf(RuntimeException.class, new ValidationError("test"));
            assertInstanceOf(RuntimeException.class, new InsufficientCreditsError());
            assertInstanceOf(RuntimeException.class, new RateLimitError());
            assertInstanceOf(RuntimeException.class, new NetworkError());
        }

        @Test
        @DisplayName("Errors can be caught as RenamedError")
        void errorsCanBeCaughtAsRenamedError() {
            try {
                throw new AuthenticationError("test");
            } catch (RenamedError e) {
                assertEquals("AUTHENTICATION_ERROR", e.getCode());
            }

            try {
                throw new ValidationError("test");
            } catch (RenamedError e) {
                assertEquals("VALIDATION_ERROR", e.getCode());
            }
        }
    }
}

File: sdks/ruby/test/errors_test.rb
# frozen_string_literal: true

require "test_helper"

class ErrorsTest < Minitest::Test
  # Base Error tests

  def test_error_default_message
    error = Renamed::Error.new

    assert_equal "An error occurred", error.message
    assert_equal "UNKNOWN_ERROR", error.code
    assert_nil error.status_code
    assert_nil error.details
  end

  def test_error_with_custom_attributes
    error = Renamed::Error.new(
      "Custom error message",
      code: "CUSTOM_CODE",
      status_code: 500,
      details: { field: "value" }
    )

    assert_equal "Custom error message", error.message
    assert_equal "CUSTOM_CODE", error.code
    assert_equal 500, error.status_code
    assert_equal({ field: "value" }, error.details)
  end

  def test_error_inherits_from_standard_error
    error = Renamed::Error.new("test")

    assert_kind_of StandardError, error
  end

  # AuthenticationError tests

  def test_authentication_error_default
    error = Renamed::AuthenticationError.new

    assert_equal "Invalid or missing API key", error.message
    assert_equal "AUTHENTICATION_ERROR", error.code
    assert_equal 401, error.status_code
  end

  def test_authentication_error_with_custom_message
    error = Renamed::AuthenticationError.new("API key is expired")

    assert_equal "API key is expired", error.message
    assert_equal "AUTHENTICATION_ERROR", error.code
    assert_equal 401, error.status_code
  end

  def test_authentication_error_inherits_from_error
    error = Renamed::AuthenticationError.new

    assert_kind_of Renamed::Error, error
  end

  # InsufficientCreditsError tests

  def test_insufficient_credits_error_default
    error = Renamed::InsufficientCreditsError.new

    assert_equal "Insufficient credits", error.message
    assert_equal "INSUFFICIENT_CREDITS", error.code
    assert_equal 402, error.status_code
  end

  def test_insufficient_credits_error_with_custom_message
    error = Renamed::InsufficientCreditsError.new("You need 10 more credits")

    assert_equal "You need 10 more credits", error.message
    assert_equal "INSUFFICIENT_CREDITS", error.code
  end

  def test_insufficient_credits_error_inherits_from_error
    error = Renamed::InsufficientCreditsError.new

    assert_kind_of Renamed::Error, error
  end

  # RateLimitError tests

  def test_rate_limit_error_default
    error = Renamed::RateLimitError.new

    assert_equal "Rate limit exceeded", error.message
    assert_equal "RATE_LIMIT_ERROR", error.code
    assert_equal 429, error.status_code
    assert_nil error.retry_after
  end

  def test_rate_limit_error_with_retry_after
    error = Renamed::RateLimitError.new("Too many requests", retry_after: 60)

    assert_equal "Too many requests", error.message
    assert_equal 60, error.retry_after
  end

  def test_rate_limit_error_inherits_from_error
    error = Renamed::RateLimitError.new

    assert_kind_of Renamed::Error, error
  end

  # ValidationError tests

  def test_validation_error
    error = Renamed::ValidationError.new("Invalid file format")

    assert_equal "Invalid file format", error.message
    assert_equal "VALIDATION_ERROR", error.code
    assert_equal 400, error.status_code
    assert_nil error.details
  end

  def test_validation_error_with_details
    details = { field: "file", errors: ["must be PDF or image"] }
    error = Renamed::ValidationError.new("Validation failed", details: details)

    assert_equal "Validation failed", error.message
    assert_equal details, error.details
  end

  def test_validation_error_inherits_from_error
    error = Renamed::ValidationError.new("test")

    assert_kind_of Renamed::Error, error
  end

  # NetworkError tests

  def test_network_error_default
    error = Renamed::NetworkError.new

    assert_equal "Network request failed", error.message
    assert_equal "NETWORK_ERROR", error.code
    assert_nil error.status_code
  end

  def test_network_error_with_custom_message
    error = Renamed::NetworkError.new("Connection refused")

    assert_equal "Connection refused", error.message
    assert_equal "NETWORK_ERROR", error.code
  end

  def test_network_error_inherits_from_error
    error = Renamed::NetworkError.new

    assert_kind_of Renamed::Error, error
  end

  # TimeoutError tests

  def test_timeout_error_default
    error = Renamed::TimeoutError.new

    assert_equal "Request timed out", error.message
    assert_equal "TIMEOUT_ERROR", error.code
    assert_nil error.status_code
  end

  def test_timeout_error_with_custom_message
    error = Renamed::TimeoutError.new("Connection timed out after 30s")

    assert_equal "Connection timed out after 30s", error.message
  end

  def test_timeout_error_inherits_from_error
    error = Renamed::TimeoutError.new

    assert_kind_of Renamed::Error, error
  end

  # JobError tests

  def test_job_error
    error = Renamed::JobError.new("Job processing failed")

    assert_equal "Job processing failed", error.message
    assert_equal "JOB_ERROR", error.code
    assert_nil error.job_id
  end

  def test_job_error_with_job_id
    error = Renamed::JobError.new("Job failed", job_id: "job_123")

    assert_equal "Job failed", error.message
    assert_equal "job_123", error.job_id
  end

  def test_job_error_inherits_from_error
    error = Renamed::JobError.new("test")

    assert_kind_of Renamed::Error, error
  end

  # error_from_http_status tests

  def test_error_from_http_status_401
    error = Renamed.error_from_http_status(401, "Unauthorized")

    assert_instance_of Renamed::AuthenticationError, error
    assert_equal "Unauthorized", error.message
  end

  def test_error_from_http_status_401_with_payload
    payload = { "error" => "Invalid API key" }
    error = Renamed.error_from_http_status(401, "Unauthorized", payload)

    assert_instance_of Renamed::AuthenticationError, error
    assert_equal "Invalid API key", error.message
  end

  def test_error_from_http_status_402
    error = Renamed.error_from_http_status(402, "Payment Required")

    assert_instance_of Renamed::InsufficientCreditsError, error
  end

  def test_error_from_http_status_400
    error = Renamed.error_from_http_status(400, "Bad Request")

    assert_instance_of Renamed::ValidationError, error
  end

  def test_error_from_http_status_422
    payload = { "error" => "Unprocessable entity", "field" => "file" }
    error = Renamed.error_from_http_status(422, "Unprocessable Entity", payload)

    assert_instance_of Renamed::ValidationError, error
    assert_equal "Unprocessable entity", error.message
    assert_equal payload, error.details
  end

  def test_error_from_http_status_429
    error = Renamed.error_from_http_status(429, "Too Many Requests")

    assert_instance_of Renamed::RateLimitError, error
  end

  def test_error_from_http_status_429_with_retry_after
    payload = { "error" => "Rate limited", "retryAfter" => 120 }
    error = Renamed.error_from_http_status(429, "Too Many Requests", payload)

    assert_instance_of Renamed::RateLimitError, error
    assert_equal 120, error.retry_after
  end

  def test_error_from_http_status_500
    error = Renamed.error_from_http_status(500, "Internal Server Error")

    assert_instance_of Renamed::Error, error
    assert_equal "Internal Server Error", error.message
    assert_equal "API_ERROR", error.code
    assert_equal 500, error.status_code
  end

  def test_error_from_http_status_503
    payload = { "error" => "Service temporarily unavailable" }
    error = Renamed.error_from_http_status(503, "Service Unavailable", payload)

    assert_instance_of Renamed::Error, error
    assert_equal "Service temporarily unavailable", error.message
    assert_equal 503, error.status_code
    assert_equal payload, error.details
  end

  def test_error_from_http_status_with_string_payload
    error = Renamed.error_from_http_status(500, "Internal Server Error", "Something went wrong")

    assert_instance_of Renamed::Error, error
    assert_equal "Internal Server Error", error.message
    assert_equal "Something went wrong", error.details
  end

  # Errors can be raised and caught

  def test_errors_can_be_raised_and_caught_by_base_class
    assert_raises(Renamed::Error) do
      raise Renamed::AuthenticationError.new
    end

    assert_raises(Renamed::Error) do
      raise Renamed::ValidationError.new("test")
    end

    assert_raises(Renamed::Error) do
      raise Renamed::RateLimitError.new
    end
  end

  def test_errors_can_be_caught_by_standard_error
    assert_raises(StandardError) do
      raise Renamed::Error.new
    end
  end
end
